third_party/amd/lib/TritonAMDGPUToLLVM/Utility.cpp

```diff
@@ -357,5 +357,26 @@ void llStore(RewriterBase &rewriter, Location loc, Value ptr, Value val,
       appendOrGetExternFuncOp(rewriter, parent, funcName, funcType);
   LLVM::createLLVMCallOp(rewriter, loc, funcOp, ValueRange({ptr, val, pred}));
 }
-
 } // namespace mlir::LLVM::AMD
+
+namespace {
+void getAllInnerForOps(scf::ForOp forOp,
+                       llvm::SetVector<scf::ForOp> &innermostForOps) {
+  bool found = false;
+  forOp.getBody()->walk([&found, &innermostForOps](scf::ForOp innerForOp) {
+    getAllInnerForOps(innerForOp, innermostForOps);
+    found = true;
+  });
+  if (!found)
+    innermostForOps.insert(forOp);
+}
+} // namespace
+
+namespace mlir::triton::AMD {
+llvm::SetVector<scf::ForOp> getAllInnerForOps(mlir::triton::FuncOp funcOp) {
+  llvm::SetVector<scf::ForOp> innermostForOps{};
+  funcOp->walk(
+      [&](scf::ForOp forOp) { ::getAllInnerForOps(forOp, innermostForOps); });
+  return innermostForOps;
+}
+} // namespace mlir::triton::AMD
```

Comment:
This implementation is more complicated than I'd expect. I think something like the following would do?

```c++
SmallVector<scf::ForOp> getLeafForOps(triton::FuncOp funcOp) {
    SmallVector<scf::ForOp> allOps;
    funcOp->walk([&](scf::ForOp forOp) { allOps.push_back(forOp); });

    SmallVector<scf::ForOp> leafOps;
    for (scf::ForOp forOp : allOps) {
      auto r = forOp->walk([](scf::ForOp) { return WalkResult::interrupt(); });
      if (!r.wasInterrupted()) leafOps.push_back(forOp);
    }
    return leafOps;
  }

```

Reply:
@antiagainst, totally agree with your implementation. It is more concise and it is doing the same. Thanks!
Done